core/domain/exp_services.py

```diff
@@ -222,6 +222,50 @@ def get_recently_published_exp_summaries(limit):
         exp_models.ExpSummaryModel.get_recently_published(limit))
 
 
+def get_story_id_linked_to_exploration(exp_id):
+    """Returns the ID of the story that the exploration is a part of or None if
+    not.
+
+    Args:
+        exp_id: str. The ID of the exploration.
+
+    Returns:
+        str|None. The ID of the story if the exploration is linked to some story
+            or None.
+    """
+    exploration_context_model = exp_models.ExplorationContextModel.get_by_id(
+        exp_id)
+    if exploration_context_model is not None:
+        return exploration_context_model.story_id
+    return None
+
+
+def update_exploration_story_link(old_exp_id, new_exp_id, story_id):
```

Comment:
This seems a bit confusing. Shouldn't this sort of thing happen precisely when a story's list of chapters is modified? It seems incorrect to have it as its own standalone function which can be called at any time -- it should be a behind-the-scenes thing that is handled along with the modification of the story chapter list, so that everything stays in sync. It should not be possible for anything else to call this in a standalone way.

It might make sense to have separate unlink_exploration_from_story_node() and link_exploration_to_story_node() functions, but each of these should make changes to both the story model and the context model so that there is no skew of the data.

Reply:
This function is indeed only called when deleting a node, linking an exploration ID to a node or deleting the story. So, changes do happen in sync. 
Since the same functionality is needed in 3 places, I made it as a function. Should I do the storage changes manually at only those locations?

Reply:
On balance, yes, let's try inlining and simplifying it, and see how that looks. I might suggest further changes after that but I'm fairly sure that this split isn't really the way to go, it should be more tightly coupled to the story-editing process and should probably be in story_services. Hopefully this gets rid of a bunch of the "if" conditions as well since in many cases what you're really doing is deleting stuff.

Reply:
Yeah, ok, will do that.

Reply:
Done, removed this function and added it to save_story.